063-topic-replication-factor-change.md

```diff
@@ -0,0 +1,271 @@
+With standalone deployment, the Cluster Operator expects two secrets that needs to be created by the user:
+
+- A secret with `cruise-control.p12` key containing the truststore in PKCS12 format, and `cruise-control.password` key containing the truststore password.
+- A secret with `cruise-control.apiToAdminUsername` key containing the API username with admin role, and `cruise-control.apiToAdminPassword` key containing the user password.
+
+### KafkaTopic CRD changes
+
+During the whole replicas change process the KafkaTopic remains in a Ready state.
+A replicas change is executed asynchronously, and takes more than one reconciliation to complete.
+
+A new optional KafkaTopicStatus subsection called `replicasChange` is used to update the user, and keep track of the replicas change state.
+At any given time, the replicas change can be pending, ongoing, or completed.
+
+- **Pending**: Not in Cruise Control's task queue (not yet sent or request error).
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:50.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 2
+    replicasChange:
+      completed: false
+    topicName: my-topic
+  ```
+
+- **Ongoing**: In Cruise Control's task queue, but execution not started, or not completed.
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:53.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 3
+    replicasChange:
+      completed: false
+      userTaskId: 1aa418ca-53ed-4b93-b0a4-58413c4fc0cb
+    topicName: my-topic
+  ```
+
+- **Completed**: Cruise Control's task execution completed (target replication factor reconciled).
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:58.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 4
+    topicName: my-topic
+  ```
+
+### Reconciliation logic
+
+At runtime, the Topic Operator watches and periodically reconciles the configuration of all managed and unpaused KafkaTopic resources.
+Among the other configurations, the reconciliation logic detects `KafkaTopic.spec.replicas` changes by comparing with the topic replication factor value in Kafka.
+
+A ReplicasChangeClient is used to isolate the logic required to call the Cruise Control REST API from the rest of the reconciliation logic.
+The ReplicasChangeClient supports sending multiple replicas changes at the same time.
+
+On every reconciliation loop, when Cruise Control integration is enabled and replicas changes are detected, the BatchingTopicController uses the ReplicasChangeClient for the following operations:
+
+1. Request pending replicas changes (uses `topic_configuration` endpoint, and returns topics with pending and ongoing changes).
+2. Check ongoing replicas changes (uses `user_tasks` endpoint, and returns topics with completed changes).
```

Comment:
Does this query all existing tasks? Or how does it find out what topics are being processed right now? I think it is the `userTaskId` that you have to query.

Reply:
It queries all existing tasks with one request. You basically pass a comma separated list of `userTaskId`s, as described in the `user_tasks` endpoint section of this proposal.